### PR TITLE
Improve support for `Body and `Alias in IR type equality checks

### PR DESCRIPTION
**Pull request by frank-emrich**
Solves #721 

There is a bit of a design question about whether or not we want a design where we rely on certain simplifications having taken place before the IR is type-checked or not. We have no mechanism in place checking that certain settings (e.g., `typecheck_ir`) may only be enabled if another setting is set (`simplify_types`).
 
Until then, this patch fixes the issue that the IR type-checker was failing unless all `Body` constructors were removed from all types beforehand. In particular, it is not necessary anymore to enable `simplify_types` for the IR type-checker to work.

### CODE REVIEW

File: core/irCheck.ml [hunk 1]
```diff
@@ -353,13 +347,17 @@ let eq_types occurrence : type_eq_context -> (Types.datatype * Types.datatype) -
       | `End, `End -> true
       | _, _ -> false
     and eq_rows (context, r1, r2) =
-      let (lfield_env, lrow_var, ldual) = remove_absent_fields_if_closed (Types.flatten_row r1) in
-      let (rfield_env, rrow_var, rdual) = remove_absent_fields_if_closed (Types.flatten_row r2) in
-      let r1 = eq_field_envs (context, lfield_env, rfield_env) in
-      let r2 = eq_row_vars (context, lrow_var, rrow_var) in
-        r1 && r2 && ldual=rdual
+      if is_toplevel_rec_row r1 || is_toplevel_rec_row r2 then
```
Comment (dhil) [inline core/irCheck.ml]: I take this as the IR typechecker cannot handle recursive rows at the moment.

**Comment by dhil** [comment]
> We have no mechanism in place checking that certain settings (e.g., `typecheck_ir`) may only be enabled if another setting is set (`simplify_types`).

I think the IR type checker should be able to handle the full type structure.

>  In particular, it is not necessary anymore to enable `simplify_types` for the IR type-checker to work.

Is there any point in retaining the type structure simplification pass? Should we just get rid of it?

**Comment by jstolarek** [comment]
I actually suggested something quite opposite to Frank, i.e. lets have passes that simplify types as much as possible so that the IR typechecker can be simpler. This would mean that the simplification passes perhaps should not be optional, so that we don't end up in a situation where we now that some given program is type incorrect but when we turn on IR typechecking we are actually checking a different program since the original program gets simplified first.

**Comment by jstolarek** [comment]
Rationale for the above is that I find it more elegant to have several simple passes, each with a clearly defined purpose, rather than having all the logic inlined in a single pass. It's just like we have multiple desugaring passes instead of just one that does all the transformations.

**Comment by dhil** [comment]
I think it is an important property to be able to run the type checker on all forms of the IR (or AST), such that it can be used a means for lightweight verification between different transformation passes.

**Comment by jstolarek** [comment]
Ok, that's a good point. I'm thinking your approach might indeed a better one.

File: core/irCheck.ml (hunk 1 above)
Comment (frank-emrich) [reply core/irCheck.ml]: Indeed, it doesn't handle recursive rows or types at the moment.

**Comment by frank-emrich** [comment]
> Is there any point in retaining the type structure simplification pass? Should we just get rid of it?

Hard to say. It can de-clutter the types when printed using the generated pp_* functions (rather than our own pretty printer, which does its own simplifications).

**Pull request merged**